rfd/0085-better-react-testing-remove-default-exports.md

```diff
@@ -0,0 +1,163 @@
+test('fetch error', () => {
+  const { getByTestId } = render(<FetchError />);
+  expect(getByTestId('Modal')).toMatchSnapshot();
+});
+```
+
+This results in all the logic in `useDesktopSession` going untested. We're only testing the UI part of the application,
+but skipping all of the business logic.
+
+Storybook isn't best suited for the way we are using it. As a result of testing our large components primarily via stories, we have
+ended up only testing our UI and nothing else.
+
+## Details
+
+### Changing how we write stories
+
+Experience tells us that Stories in Storybook are used most effectively for documenting how to use reusable components. We should therefore
+restrict usage of Storybook mostly to our shared component library, and test our fullscreen components another way.
+
+It's still okay to create our larger components in Storybook (if the developer wants to) but we should not be using these stories in our Jest tests.
+
+It might be worthwhile looking into something like [Chromatic](https://www.chromatic.com/) for visual UI testing.
+
+### Changing how we write tests
+
+We should be testing all the different state possibilities in Jest and React Testing Library, mocking any data or
+network requests to cause the component to render into the state we want. This ensures that the internal logic
+of our components (generally the most critical and difficult-to-get-right aspect of frontend code) gets tested,
+as well as the final UI state.
+
+Tests should check for elements that exist for the specific state being tested, i.e. an error message when the data has
+failed to load. Elements should be clicked on or events triggered in order to change the state of the component.
+
+By writing tests as described above, snapshot tests typically become redundant and therefore unnecessary. Besides the problem of skipping the business logic, snapshots consistently break for non-bugs, such as when styles are updated, and
+generally don't provide much if any security around behaviour or appearance. Further reading -
+https://medium.com/@sapegin/whats-wrong-with-snapshot-tests-37fbe20dfe8e.
+
+### Changing how we export components
+
+Default exports should be avoided except when absolutely necessary (for example, with
+`React.Lazy`), as they degrade the developer experience
+for no discernible benefit. For more fully fleshed out reasoning, refer to the articles below:
+- https://blog.neufund.org/why-we-have-banned-default-exports-and-you-should-do-the-same-d51fdc2cf2ad
+- https://rajeshnaroth.medium.com/avoid-es6-default-exports-a24142978a7a
+- https://blog.piotrnalepa.pl/2020/06/26/default-exports-vs-named-exports/
+- https://ilikekillnerds.com/2019/08/default-exports-bad/
+
+When needing to export something as default, we should export it like so:
+
+```typescript jsx
+function SomeComponent() {
+  return (
+    <div>
+      hello!
+    </div>
+  );
+}
+
+export { SomeComponent as default };
+```
+
+This provides a clearer, more explicit export.
```

Comment:
In what way is this clearer and more explicit than `export default function SomeComponent`? If anything I'd say that it makes it less clear because you have to hunt for that export statement. 😅 Especially if the export statement is added at the end of a long component body.

At first I thought that this way of using `export` has some additional benefits in terms of TypeScript usage or something but it doesn't appear so?

Reply:
Well, it should be right below the function. I'm not really fussed on this, it would just keep a consistent export of defaults (keeps it consistent with `export { act };`)

Reply:
> Well, it should be right below the function. I'm not really fussed on this, it would just keep a consistent export of defaults…

I ultimately don't feel that strong about it either way too, I suppose my issue is that when we write something down we should have a solid argumentation for it. Here I just couldn't see how this is clearer over the other form.

Now, keeping a consistent export of defaults – that's a good argument. So perhaps we could say that we want to use this form for consistency, not because it's objectively better.

> …(keeps it consistent with `export { act };`)

What are you referring to here? I've always did pretty much all of my exports by adding `export` to where the given variable was defined and I'm not sure if that goes against what you're saying here.